File: tests/fixtures/funding_safety.py
#!/usr/bin/env python3
"""Proof suite for the funding-identity invariant (0.4.1 S5a): an included-only
profile never launches when its adapter observes funding other than what the
user authorized, and a refusal stays until setup re-authorizes the profile.

Codex scenarios run with `codex_probe: false`, so the capacity path never
probes the account and every refusal here comes from the adapter preflight.
Never invokes a model provider; the provider executables are fixtures."""
import json
import sqlite3
import sys

sys.dont_write_bytecode = True
from provider_fixture import Fixture


def run(f):
    return f.command('run', f.source, '--task', 'Add tests in src/lib.rs',
                     '--allow-unsafe-local', '--json', check=False)


def refused(f, output, reason):
    stderr = output.stderr.decode()
    assert output.returncode != 0, (output.stdout, stderr)
    assert f.count() == 0, 'a refused launch must not spawn the agent'
    assert reason in stderr, (reason, stderr)


def launched(f, output, count=1):
    assert output.returncode == 0, (output.stdout, output.stderr)
    assert f.count() == count, f.count()


def refusals(f):
    with sqlite3.connect(f.state / 'dispatch.db') as db:
        return db.execute('SELECT authorization_revision, reason FROM funding_refusals').fetchall()


def reauthorize(f):
    path = f.state / 'resources.yml'
    path.write_text(path.read_text().replace('authorization_revision: 1', 'authorization_revision: 2'))


CODEX = {
    # C1-C5 and the unknown-identity decision, one fixture each.
    'auth': ({'type': 'apiKey', 'planType': 'plus', 'email': 'fixture@example.invalid'}, None,
             'authentication changed to apiKey'),
    'credits': (None, {'rateLimitsByLimitId': {'codex': {'credits': {'hasCredits': True}}}},
                'paid credits are now available'),
    'tier': (None, {'serviceTier': 'priority', 'rateLimitsByLimitId': {}},
             'service tier changed to priority'),
    'plan': ({'type': 'chatgpt', 'planType': 'pro', 'email': 'fixture@example.invalid'}, None,
             'funding plan changed to pro'),
    'account': ({'type': 'chatgpt', 'planType': 'plus', 'email': 'other@example.invalid'}, None,
                'the Codex account changed'),
    'unknown': ({'type': 'chatgpt', 'planType': 'plus'}, None,
                'could not be observed'),
}


def scenario(binary, name):
    f = Fixture(binary)
    try:
        if name == 'codex_authorized':
            launched(f, run(f))
            assert refusals(f) == []
        elif name in CODEX:
            account, limits, reason = CODEX[name]
            if account:
                (f.root / 'codex-account.json').write_text(json.dumps(account))
            if limits:
                (f.root / 'codex-limits.json').write_text(json.dumps(limits))
            refused(f, run(f), reason)
            assert [r for (r, _) in refusals(f)] == [1], refusals(f)
        elif name == 'codex_unreadable':
            (f.root / 'codex-probe').write_text('exit')
            refused(f, run(f), 'could not be read')
        elif name == 'codex_missing_evidence':
            path = f.state / 'resources.yml'
            path.write_text(''.join(l for l in path.read_text().splitlines(True) if 'codex_account' not in l))
            refused(f, run(f), 'Codex account evidence missing')
        elif name == 'codex_spawn_boundary':
            # Selection's preflight sees the authorized account; the preflight
            # at the spawn boundary sees another. Nothing is spawned and the
            # refusal is recorded, so the next run is refused at selection.
            (f.root / 'codex-switch-at').write_text('2')
            refused(f, run(f), 'the Codex account changed')
            assert [r for (r, _) in refusals(f)] == [1], refusals(f)
            (f.root / 'codex-switch-at').unlink()
            refused(f, run(f), 'was refused')
        elif name == 'codex_profile_changed_after_selection':
            # C9: the bound profile changes in resources.yml after selection.
            (f.root / 'codex-mutate-resources-at').write_text('1')
            refused(f, run(f), 'resource configuration changed after selection')
        elif name in ('codex_sticky', 'claude_sticky'):
            if name == 'codex_sticky':
                switch = f.root / 'codex-account.json'
                switch.write_text(json.dumps({'type': 'chatgpt', 'planType': 'plus',
                                              'email': 'other@example.invalid'}))
                refused(f, run(f), 'the Codex account changed')
                switch.unlink()
            else:
                auth = f.root / 'auth.json'
                original = auth.read_text()
                value = json.loads(original)
                value['email'] = 'other@example.invalid'
                auth.write_text(json.dumps(value))
                refused(f, run(f), 'account changed')
                auth.write_text(original)
            # The account is back, but the refusal holds until re-authorization.
            refused(f, run(f), 'was refused')
            # A new authorization revision is not refused by the preflight's
            # record. (Launching after re-authorization is asserted once the
            # capacity path, which keeps its own stale rejection, is removed.)
            reauthorize(f)
            output = run(f)
            assert b'was refused' not in output.stderr, output.stderr
            assert [r for (r, _) in refusals(f)] == [1], refusals(f)
        else:
            raise SystemExit(f'unknown scenario {name}')
    finally:
        f.cleanup()


if __name__ == '__main__':
    scenario(sys.argv[1], sys.argv[2])

File: tests/funding_safety.rs
//! Funding-identity proof suite (0.4.1 S5a; the gate for deleting capacity and
//! admission in S6b, which must leave this file unchanged). Each case runs the
//! real binary against fixture provider executables; see
//! `tests/fixtures/funding_safety.py` and `docs/plan-0.4.1-pruning.md`.
#![cfg(unix)]
use std::process::Command;

fn scenario(provider: &str, name: &str) {
    let output = Command::new("python3")
        .arg(format!(
            "{}/tests/fixtures/funding_safety.py",
            env!("CARGO_MANIFEST_DIR")
        ))
        .arg(assert_cmd::cargo_bin!("dispatch"))
        .arg(name)
        .env("DISPATCH_FIXTURE_PROVIDER", provider)
        .env("PYTHONDONTWRITEBYTECODE", "1")
        .output()
        .expect("Python standard-library fixture");
    assert!(
        output.status.success(),
        "{provider} {name}\n{}\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );
}

macro_rules! cases {
    ($($test:ident => ($provider:literal, $name:literal)),* $(,)?) => {
        $(#[test] fn $test() { scenario($provider, $name); })*
    };
}

cases! {
    codex_authorized_subscription_launches => ("codex", "codex_authorized"),
    codex_other_authentication_is_refused => ("codex", "auth"),
    codex_available_paid_credits_are_refused => ("codex", "credits"),
    codex_non_standard_service_tier_is_refused => ("codex", "tier"),
    codex_other_plan_is_refused => ("codex", "plan"),
    codex_other_account_is_refused => ("codex", "account"),
    codex_unobservable_identity_is_refused => ("codex", "unknown"),
    codex_unreadable_account_is_refused => ("codex", "codex_unreadable"),
    codex_missing_account_evidence_is_refused => ("codex", "codex_missing_evidence"),
    codex_change_at_the_spawn_boundary_is_refused_and_recorded => ("codex", "codex_spawn_boundary"),
    codex_profile_changed_after_selection_is_refused => ("codex", "codex_profile_changed_after_selection"),
    codex_refusal_is_sticky_until_reauthorized => ("codex", "codex_sticky"),
    claude_refusal_is_sticky_until_reauthorized => ("claude", "claude_sticky"),
}
